### PR TITLE
Add posgresql support

### PR DESCRIPTION
**Pull request by MartinSchmidt**
- Added dbup and postgresql implementaion for the registry
- Added migration job to helm chart
- Added CloudnativePG support for helm chart
- Added support for unbalanced Merkle trees and blocks
- BlockFinalizer now only is configured with duration and not blocksize
- Renamed batches to blocks to better highlight what it is.
- Fixed code smell - Removed implicit usings 
- Concordium implementation updated to new library 

Known issues
is broken since their api changes, new version will come in seperate PR
- Documentation has not been updated, will be done in seperate PR

### CODE REVIEW

File: src/ProjectOrigin.VerifiableEventStore/Extensions/IEnumerableMerkleExtension.cs [hunk 1]
```diff
@@ -7,75 +7,85 @@ namespace ProjectOrigin.VerifiableEventStore.Extensions;
 
 public static class IEnumerableMerkleExtension
 {
-    public static byte[] CalculateMerkleRoot<T>(this IEnumerable<T> events, Func<T, byte[]> selector)
+    public static byte[] CalculateMerkleRoot<T>(this IList<T> events, Func<T, byte[]> selector)
```
Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Extensions/IEnumerableMerkleExtension.cs:10]: Why is this changed to `IList`? It looks like that it can be an `IEnumerable` and then it will match the name of the static class.

File: src/ProjectOrigin.VerifiableEventStore/Models/BlockHash.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        if (right is null)
+            return false;
+
+        if (Data == null || right.Data == null)
+        {
+            return Data == right.Data;
+        }
+        return Data.SequenceEqual(right.Data);
+    }
+    public override int GetHashCode()
```
Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Models/BlockHash.cs:22]:
```suggestion

    public override int GetHashCode()
```

File: src/ProjectOrigin.VerifiableEventStore/Models/TransactionHash.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        if (right is null)
+            return false;
+
+        if (Data == null || right.Data == null)
+        {
+            return Data == right.Data;
+        }
+        return Data.SequenceEqual(right.Data);
+    }
+    public override int GetHashCode()
```
Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Models/TransactionHash.cs:18]:
```suggestion

    public override int GetHashCode()
```

File: src/ProjectOrigin.VerifiableEventStore/Services/BlockFinalizer/BlockFinalizerBackgroundService.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            _serviceProvider.GetRequiredService<ILogger<BlockFinalizerJob>>(),
+            _serviceProvider.GetRequiredService<IBlockPublisher>(),
+            _serviceProvider.GetRequiredService<ITransactionRepository>(),
+            _serviceProvider.GetRequiredService<ITransactionStatusService>()
+            );
+
+        using var timer = new PeriodicTimer(_blockInterval);
+
+        while (!stoppingToken.IsCancellationRequested &&
+               await timer.WaitForNextTickAsync(stoppingToken))
```
Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Services/BlockFinalizer/BlockFinalizerBackgroundService.cs:40]: Is there a chance that when the tick occurs on the timer, the stoppingToken has been cancelled, but the code will continue?

File: src/ProjectOrigin.VerifiableEventStore/Services/BlockPublisher/Log/LogPublisher.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
+using System;
+using System.Threading.Tasks;
+using Google.Protobuf;
+using Microsoft.Extensions.Logging;
+using ProjectOrigin.VerifiableEventStore.Models;
+
+namespace ProjectOrigin.VerifiableEventStore.Services.BlockPublisher.Log;
+
+/// The implementation can be used locally to see the output in the log instead of
+/// having to send it to
```
Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Services/BlockPublisher/Log/LogPublisher.cs:10]: "having to send it to"... what?

File: src/ProjectOrigin.VerifiableEventStore/Services/Repository/ITransactionRepository.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+using ProjectOrigin.VerifiableEventStore.Models;
+
+namespace ProjectOrigin.VerifiableEventStore.Services.Repository;
+
+public interface ITransactionRepository
+{
+    Task Store(StreamTransaction @event);
+
+    Task<ImmutableLog.V1.Block?> GetBlockFromTransactionHash(TransactionHash transactionHash);
+    Task<TransactionStatus> GetTransactionStatus(TransactionHash transactionHash);
```
Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Services/Repository/ITransactionRepository.cs:13]: Should this be 

```suggestion
    Task<TransactionStatus?> GetTransactionStatus(TransactionHash transactionHash);
```

Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Services/Repository/ITransactionRepository.cs:13]: The naming of these two differ a bit. I suggest to remove the `FromTransactionHash` part of `GetBlockFromTransactionHash`

File: src/ProjectOrigin.VerifiableEventStore/Services/Repository/ITransactionRepository.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    Task<IList<StreamTransaction>> GetStreamTransactionsForBlock(BlockHash blockHash);
+    Task<IList<StreamTransaction>> GetStreamTransactionsForStream(Guid streamId);
+
+    Task<NewBlock?> CreateNextBlock();
+    Task FinalizeBlock(BlockHash hash, ImmutableLog.V1.BlockPublication publication);
+
+    public record NewBlock(ImmutableLog.V1.BlockHeader Header, IList<TransactionHash> TransactionHashes);
```
Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Services/Repository/ITransactionRepository.cs:20]: Is it on purpose that this record is defined within the interface?

File: src/ProjectOrigin.VerifiableEventStore/Services/Repository/InMemory/InMemoryRepository.cs [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+using ProjectOrigin.VerifiableEventStore.Extensions;
+using ProjectOrigin.VerifiableEventStore.Models;
+using ProjectOrigin.VerifiableEventStore.Services.Repository;
+
+namespace ProjectOrigin.VerifiableEventStore.Services.EventStore.InMemory;
+
+public class InMemoryRepository : ITransactionRepository
+{
+    private readonly BlockSizeCalculator _blockSizeCalculator;
+    private object lockObject = new();
```
Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Services/Repository/InMemory/InMemoryRepository.cs:18]:
```suggestion
    private object _lockObject = new();
```

File: src/ProjectOrigin.VerifiableEventStore/Services/Repository/Postgres/PostgresqlRepository.cs [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    private Task<long?> GetTransactionId(IDbConnection connection, TransactionHash transactionHash)
+    {
+        return connection.QuerySingleOrDefaultAsync<long?>(
+              "SELECT id FROM transactions WHERE transaction_hash = @transactionHash",
+              new { transactionHash = transactionHash.Data });
+    }
+
+    private record DBBlock
```
Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Services/Repository/Postgres/PostgresqlRepository.cs:213]: It is a bit strange that `DBBlock` is private and at the same time is something that can be fetched from the database.

File: src/ProjectOrigin.VerifiableEventStore/Services/TransactionStatusCache/TransactionStatusService.cs [hunk 10]
```diff
@@ -4,44 +4,54 @@
...
 
 public class TransactionStatusService : ITransactionStatusService
 {
     private static readonly TimeSpan CacheTime = TimeSpan.FromMinutes(60);
+    private static readonly TimeSpan UnknownCacheTime = TimeSpan.FromMinutes(5);
+
     private ILogger<TransactionStatusService> _logger;
     private IDistributedCache _cache;
-    private IEventStore _eventStore;
+    private ITransactionRepository _eventStore;
```
Comment (duizer) [inline src/ProjectOrigin.VerifiableEventStore/Services/TransactionStatusCache/TransactionStatusService.cs:18]:
```suggestion
    private ITransactionRepository _transactionRepository;
```

File: charts/project-origin-registry/templates/registry-deployment.yaml [hunk 11]
```diff
@@ -1,4 +1,7 @@
----
+{{- if not (or (and (not (.Values.persistance.cloudNativePG.enabled)) .Values.persistance.inMemory.enabled) (and .Values.persistance.cloudNativePG.enabled (not (.Values.persistance.inMemory.enabled)))) }}
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply charts/project-origin-registry/templates/registry-deployment.yaml:1]: xor is not a valid command, so implemented it using and or and not. ;)

So basically 1 and only one of the two may be enabled

File: src/ProjectOrigin.Registry.Server/Services/VerifyTransactionConsumer.cs [hunk 12]
```diff
@@ -10,21 +10,21 @@
...
     private TransactionProcessorOptions _options;
-    private IEventStore _eventStore;
+    private ITransactionRepository _eventStore;
     private ITransactionDispatcher _verifier;
     private ITransactionStatusService _transactionStatusService;
     private ILogger<VerifyTransactionConsumer> _logger;
 
     public VerifyTransactionConsumer(IOptions<TransactionProcessorOptions> options,
-                                IEventStore localEventStore,
+                                ITransactionRepository localEventStore,
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.Registry.Server/Services/VerifyTransactionConsumer.cs]: done

File: src/ProjectOrigin.VerifiableEventStore.Tests/AbstractEventStoreTests.cs [hunk 13]
```diff
@@ -1,158 +1,277 @@
...
+using ProjectOrigin.VerifiableEventStore.Extensions;
 using ProjectOrigin.VerifiableEventStore.Models;
 using ProjectOrigin.VerifiableEventStore.Services.EventStore;
+using ProjectOrigin.VerifiableEventStore.Services.Repository;
+using Xunit;
 
 namespace ProjectOrigin.VerifiableEventStore.Tests;
 
-public abstract class AbstractEventStoreTests<T> where T : IEventStore
+public abstract class AbstractEventStoreTests<T> where T : ITransactionRepository
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore.Tests/AbstractEventStoreTests.cs:17]: done

File: src/ProjectOrigin.VerifiableEventStore.Tests/PostgresqlEventStoreTests.cs [hunk 14]
```diff
@@ -1,21 +1,35 @@
...
 
-// namespace ProjectOrigin.VerifiableEventStore.Tests;
+namespace ProjectOrigin.VerifiableEventStore.Tests;
 
-// public class PostgresqlEventStoreTests : AbstractEventStoreTests<PostgresqlEventStore>, IClassFixture<DatabaseFixture>
-// {
-//     private PostgresqlEventStore _eventStore;
+public class PostgresqlEventStoreTests : AbstractEventStoreTests<PostgresqlRepository>, IClassFixture<PostgresDatabaseFixture>, IAsyncLifetime
+{
+    private readonly PostgresqlRepository _eventStore;
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore.Tests/PostgresqlEventStoreTests.cs]: done

File: src/ProjectOrigin.VerifiableEventStore/Services/TransactionStatusCache/TransactionStatusService.cs (hunk 10 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Services/TransactionStatusCache/TransactionStatusService.cs]: done

File: src/ProjectOrigin.VerifiableEventStore/Services/Repository/Postgres/PostgresqlRepository.cs (hunk 9 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Services/Repository/Postgres/PostgresqlRepository.cs]: changed name

File: src/ProjectOrigin.VerifiableEventStore/Services/Repository/InMemory/InMemoryRepository.cs (hunk 8 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Services/Repository/InMemory/InMemoryRepository.cs]: done

File: src/ProjectOrigin.VerifiableEventStore/Services/Repository/InMemory/InMemoryRepository.cs [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    private List<StreamTransaction> _events = new();
+    private ConcurrentDictionary<BlockHash, BlockRecord> _blocks = new();
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Services/Repository/InMemory/InMemoryRepository.cs]: done

File: src/ProjectOrigin.VerifiableEventStore/Extensions/IEnumerableMerkleExtension.cs (hunk 1 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Extensions/IEnumerableMerkleExtension.cs:10]: IEnumerable interface does not guarantee order, but the order is important in this regard, therefore forcing people using the function to send a IList forces them to think about the issue.

File: src/ProjectOrigin.Registry.Server/Services/RegistryService.cs [hunk 16]
```diff
@@ -33,11 +34,11 @@ public override async Task<SubmitTransactionResponse> SendTransactions(SendTrans
         foreach (var transaction in request.Transactions)
         {
             var message = VerifyTransaction.Create(transaction);
-            var transactionHash = Convert.ToBase64String(SHA256.HashData(transaction.ToByteArray()));
+            var transactionHash = new TransactionHash(SHA256.HashData(transaction.ToByteArray()));
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.Registry.Server/Services/RegistryService.cs]: done

File: src/ProjectOrigin.Registry.Server/Services/RegistryService.cs [hunk 17]
```diff
@@ -17,11 +18,11 @@ public class RegistryService : V1.RegistryService.RegistryServiceBase
     public static Meter Meter = new("Registry.RegistryService");
     public static Counter<long> TransactionsSubmitted = Meter.CreateCounter<long>("TransactionsSubmitted");
 
-    private IEventStore _eventStore;
+    private ITransactionRepository _eventStore;
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.Registry.Server/Services/RegistryService.cs]: done

File: src/ProjectOrigin.Registry.Server/Services/VerifyTransactionConsumer.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.Registry.Server/Services/VerifyTransactionConsumer.cs]: done

File: src/ProjectOrigin.Registry.Server/Services/VerifyTransactionConsumer.cs [hunk 18]
```diff
@@ -10,21 +10,21 @@
...
+using ProjectOrigin.VerifiableEventStore.Services.Repository;
 using ProjectOrigin.VerifiableEventStore.Services.TransactionStatusCache;
 
 namespace ProjectOrigin.Registry.Server.Services;
 
 public class VerifyTransactionConsumer : IConsumer<VerifyTransaction>
 {
     private TransactionProcessorOptions _options;
-    private IEventStore _eventStore;
+    private ITransactionRepository _eventStore;
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.Registry.Server/Services/VerifyTransactionConsumer.cs]: done

File: src/ProjectOrigin.VerifiableEventStore/Models/TransactionHash.cs (hunk 3 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Models/TransactionHash.cs:18]: done

File: src/ProjectOrigin.VerifiableEventStore/Services/BlockFinalizer/BlockFinalizerBackgroundService.cs (hunk 4 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Services/BlockFinalizer/BlockFinalizerBackgroundService.cs:40]: I don't think so, if the stoppingToken is canceled, then it will break the while loop, and the stoppingtoken is also send to the waitForNext, so if cancelled, then it should also stop it.

File: src/ProjectOrigin.VerifiableEventStore/Services/BlockPublisher/Log/LogPublisher.cs (hunk 5 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Services/BlockPublisher/Log/LogPublisher.cs:10]: done

File: src/ProjectOrigin.VerifiableEventStore/Services/EventProver/EventProverService.cs [hunk 19]
```diff
@@ -1,33 +1,34 @@
...
-using ProjectOrigin.VerifiableEventStore.Services.EventStore;
+using ProjectOrigin.VerifiableEventStore.Models;
+using ProjectOrigin.VerifiableEventStore.Services.Repository;
 
 namespace ProjectOrigin.VerifiableEventStore.Services.EventProver;
 
 public class EventProverService : IEventProver
 {
-    private IEventStore _eventStore;
+    private ITransactionRepository _eventStore;
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Services/EventProver/EventProverService.cs]: done

File: src/ProjectOrigin.VerifiableEventStore/Services/Repository/ITransactionRepository.cs (hunk 6 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Services/Repository/ITransactionRepository.cs:13]: Actually no, the TransactionStatus is the following enum

```csharp
public enum TransactionStatus
{
    Unknown = 0,
    Pending = 1,
    Failed = 2,
    Committed = 3
}
```

so it simply returns Unknown = 0

Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Services/Repository/ITransactionRepository.cs:13]: Fixed

File: src/ProjectOrigin.VerifiableEventStore/Services/Repository/ITransactionRepository.cs (hunk 7 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.VerifiableEventStore/Services/Repository/ITransactionRepository.cs:20]: kinda, in the first iteration i returned a tuple of the two variables, but needed to be able to return a null reference when no new block could be created because of no transactions to put in the block.. But it could also have been achived using a try pattern

```csharp
bool TryCreateNextBlock(out ImmutableLog.V1.BlockHeader Header, out IList<TransactionHash> TransactionHashes);
```

This might actually make it more readable, but it always depends on people preference of the try methods with multiple out parameters.

File: charts/project-origin-registry/templates/registry-deployment.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (duizer) [reply charts/project-origin-registry/templates/registry-deployment.yaml:1]: You are right - it should be an XOR operation.

File: src/ProjectOrigin.VerifiableEventStore/Services/BlockFinalizer/BlockFinalizerBackgroundService.cs (hunk 4 above)
Comment (duizer) [reply src/ProjectOrigin.VerifiableEventStore/Services/BlockFinalizer/BlockFinalizerBackgroundService.cs:40]: I think `!stoppingToken.IsCancellationRequested` is executed first, then `await timer.WaitForNextTickAsync(stoppingToken)` will be executed. If `await timer.WaitForNextTickAsync(stoppingToken)` stops due to the stoppingToken being cancelled, then `!stoppingToken.IsCancellationRequested` is not executed again. 
But I am not sure about the exact behaviour.

**Review by duizer — APPROVED** [review]
